<table align="left" width=100%>
    <tr>
        <td width="20%">
            <img src="faculty.png">
        </td>
        <td>
            <div align="center">
                <font color="#21618C" size=8px>
                  <b> Assignment Questions </b>
                </font>
            </div>
        </td>
    </tr>
</table>

<table align="left" width=100%>
    <tr>
        <td>
            <div align="center">
                <font color="#21618C" size=5px>
                  <b>Group-16</b>
                </font>
            </div>
        </td>
    </tr>
    <tr>
        <td>
            <div align="left">
                <font color="#21618C" size=5px>
                  <b> V S Shabarish </b>
                </font>
            </div>
        </td>
    </tr>
    <tr>
        <td>
            <div align="left">
                <font color="#21618C" size=5px>
                  <b> Tania Khan </b>
                </font>
            </div>
        </td>
    </tr>
    <tr>
        <td>
            <div align="left">
                <font color="#21618C" size=5px>
                  <b> Vinod A </b>
                </font>
            </div>
        </td>
    </tr>
</table>

Total Marks:  30 Marks


### About the data set (COMPANY_Reviews)

The dataset contains Text reviews. The aim of this assignment is to classsify the review texts into two categories. 

Attribute information: 
 
Dataset has two columns:
Column2: Review Text
Column1: Two Classes. Positive and negative.

## Table of Content

1. **[Import Libraries](#lib)**
2. **[Load the Data  (5 Marks)](#prep)**
    - 2.1 - **[Read the Data   (1 Mark)](#read)**
    - 2.2 - **[Dispay the first ten rows (1 Mark) ](#dtype)**
    - 2.3 - **[Display the information about the data (1 Mark)](#drop)**
    - 2.4 - **[Encode the sentiment column values as 1 or 0 (2 Marks)](#dummy)**
3.  **[Data preprocessing  (7 Marks)](#prep)**
    - 3.1 - **[Remove special characters and html tags (2 Marks)](#read)**
    - 3.2 - **[Convert reviews into lowercase (1 Mark)](#dtype)**
    - 3.3 - **[Removal of stop words (2 Marks)](#drop)**
    - 3.4 - **[Apply stemming (2 Marks)](#dummy)**
4.**[Convert review text into feature vector and classification (18 Marks)](#prep)**
    - 4.1 - **[create BOW model (2 Marks)](#cut_off)**
    - 4.2 - **[Training data and Test data -split  (1 Mark)](#dtype)**
    - 4.3 - **[Define the Classification model (any one like NB, SVM or Random forest) train the model  (3 Marks) ](#drop)**
    - 4.4 - **[Classification metrics analysis (2 Marks)](#dummy)**
    - 4.5 - **[predict the class for your own review (2 Marks)](#dummy)**

    - 4.6 - **[create feature vector using tf-idf  (2 Marks)](#cut_off)**
    - 4.7 - **[Define the Classification model (any one like NB, SVM or Random forest) train the model (2 Marks)](#drop)**
    - 4.8 - **[Classification metrics analysis (2 Marks)](#dummy)**
    - 4.9 - **[predict the class for your own review (2 Marks) ](#dummy)**

        

<a id="lib"></a>
# 1. Import Libraries

**Let us import the required libraries.**

In [1]:
import pandas as pd
import re
from bs4 import BeautifulSoup
import spacy
import nltk
from nltk.stem import PorterStemmer
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer

from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import MultinomialNB
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

import warnings
warnings.filterwarnings("ignore")

C:\Users\vssha\anaconda3\lib\site-packages\scipy\__init__.py:155: UserWarning: A NumPy version >=1.18.5 and <1.25.0 is required for this version of SciPy (detected version 1.25.2
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


2. Load the Data

2.1 Read the data

In [2]:
company_review = pd.read_csv("sentiment_analysis.csv",encoding='ANSI')

<font color='blue'>We are using a encoding to read the document without any errors.</font> 


2.2 Dispay the first ten rows

In [3]:
company_review.head(10)

Sentiment                                             Review
0  negative  The international electronic industry company ...
1  positive  With the new production plant the company woul...
2  positive  According to the company 's updated strategy f...
3  positive  FINANCING OF ASPOCOMP 'S GROWTH Aspocomp is ag...
4  positive  For the last quarter of 2010 , Componenta 's n...
5  positive  In the third quarter of 2010 , net sales incre...
6  positive  Operating profit rose to EUR 13.1 mn from EUR ...
7  positive  Operating profit totalled EUR 21.1 mn , up fro...
8  positive  TeliaSonera TLSN said the offer is in line wit...
9  positive  STORA ENSO , NORSKE SKOG , M-REAL , UPM-KYMMEN...

<font color='blue'>From the above we can see that this is a review and sentiment analysis of a company or a industry and most of them focus about the growth.It has positive and negative sentiments.</font> 

2.3 Display the information about the data

In [4]:
company_review.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1967 entries, 0 to 1966
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Sentiment  1967 non-null   object
 1   Review     1967 non-null   object
dtypes: object(2)
memory usage: 30.9+ KB


<font color='blue'>There are 1967 records in the table and there are no null values in both sentiment and review.</font> 

In [5]:
company_review.describe()

Sentiment                                             Review
count       1967                                               1967
unique         2                                               1967
top     positive  The international electronic industry company ...
freq        1363                                                  1

2.4   Encode the sentiment column values as 1 or 0

In [6]:
company_review["Sentiment"].replace({"negative":1, "positive":0}, inplace=True)

<font color='blue'>here we are label encoding positive and negative review negative becomes 1 and positive is 0.</font> 

In [7]:
company_review.head()

Sentiment                                             Review
0          1  The international electronic industry company ...
1          0  With the new production plant the company woul...
2          0  According to the company 's updated strategy f...
3          0  FINANCING OF ASPOCOMP 'S GROWTH Aspocomp is ag...
4          0  For the last quarter of 2010 , Componenta 's n...

3.  Data preprocessing

3.1 Remove special characters and html tags

In [8]:
def remove_htmltags_specialcharacters(text):
    soup = BeautifulSoup(text, "html.parser")
    clean_text = soup.get_text()
    clean_text = re.sub(r'[^A-Za-z0-9\s]+', '', clean_text)
    return clean_text

In [9]:
company_review['Cleaned_Review'] = company_review['Review'].apply(remove_htmltags_specialcharacters)
company_review.head()

Sentiment                                             Review  \
0          1  The international electronic industry company ...   
1          0  With the new production plant the company woul...   
2          0  According to the company 's updated strategy f...   
3          0  FINANCING OF ASPOCOMP 'S GROWTH Aspocomp is ag...   
4          0  For the last quarter of 2010 , Componenta 's n...   

                                      Cleaned_Review  
0  The international electronic industry company ...  
1  With the new production plant the company woul...  
2  According to the company s updated strategy fo...  
3  FINANCING OF ASPOCOMP S GROWTH Aspocomp is agg...  
4  For the last quarter of 2010  Componenta s net...

<font color='blue'>The reason to hold ansi encoding is to retain special characters that can be now cleaned by this process.</font> 

3.2 - Convert reviews into lowercase

In [10]:
def lower_text(text):
    lower_text = text.lower()
    return lower_text

In [11]:
company_review['Cleaned_Review'] = company_review['Cleaned_Review'].apply(lower_text)
company_review.head()

Sentiment                                             Review  \
0          1  The international electronic industry company ...   
1          0  With the new production plant the company woul...   
2          0  According to the company 's updated strategy f...   
3          0  FINANCING OF ASPOCOMP 'S GROWTH Aspocomp is ag...   
4          0  For the last quarter of 2010 , Componenta 's n...   

                                      Cleaned_Review  
0  the international electronic industry company ...  
1  with the new production plant the company woul...  
2  according to the company s updated strategy fo...  
3  financing of aspocomp s growth aspocomp is agg...  
4  for the last quarter of 2010  componenta s net...

3.3 - Removal of stop words

In [12]:
nlp = spacy.load("en_core_web_lg")

def remove_stopwords(text):
    doc = nlp(text)
    filtered_text = [token.text for token in doc if not token.is_stop]
    return ' '.join(filtered_text)

In [13]:
company_review['Cleaned_Review'] = company_review['Cleaned_Review'].apply(remove_stopwords)
company_review.head()

Sentiment                                             Review  \
0          1  The international electronic industry company ...   
1          0  With the new production plant the company woul...   
2          0  According to the company 's updated strategy f...   
3          0  FINANCING OF ASPOCOMP 'S GROWTH Aspocomp is ag...   
4          0  For the last quarter of 2010 , Componenta 's n...   

                                      Cleaned_Review  
0  international electronic industry company elco...  
1  new production plant company increase capacity...  
2  according company s updated strategy years 200...  
3  financing aspocomp s growth aspocomp aggressiv...  
4  quarter 2010   componenta s net sales doubled ...

<font color='blue'>we are now removing all the stop word present in the cleaned dataset and overwriting the same so that we can improve our prediction rate .</font> 

3.4 - Apply stemming

In [14]:
def stem_text(text):
    stemmer = PorterStemmer()
    tokens = nltk.word_tokenize(text)
    stemmed_tokens = [stemmer.stem(token) for token in tokens]
    return ' '.join(stemmed_tokens)

In [15]:
company_review['Cleaned_Review'] = company_review['Cleaned_Review'].apply(stem_text)
company_review.head()

Sentiment                                             Review  \
0          1  The international electronic industry company ...   
1          0  With the new production plant the company woul...   
2          0  According to the company 's updated strategy f...   
3          0  FINANCING OF ASPOCOMP 'S GROWTH Aspocomp is ag...   
4          0  For the last quarter of 2010 , Componenta 's n...   

                                      Cleaned_Review  
0  intern electron industri compani elcoteq laid ...  
1  new product plant compani increas capac meet e...  
2  accord compani s updat strategi year 20092012 ...  
3  financ aspocomp s growth aspocomp aggress purs...  
4  quarter 2010 componenta s net sale doubl eur13...

<font color='blue'>Stemming cutts the word to its root word and makes the predictablity more easier it becomes easier to implement Bag of words and TF-IDF for making train test split.</font> <font color='red'> in our case we have not done lemmatization but it is also important to do lemmatization</font>

In [16]:
def preprocess_text(text):
    # Remove HTML tags and Special characters
    clean_text = remove_htmltags_specialcharacters(text)
   
    # Convert to lowercase
    clean_text = lower_text(clean_text)
    
    # Remove Stop words
    clean_text = remove_stopwords(clean_text)
    
    # Apply Stemming
    clean_text = stem_text(clean_text)
    
    return clean_text

4.  Convert review text into feature vector and classification

4.1 - create BOW model 

In [17]:
cv = CountVectorizer()
X = cv.fit_transform(company_review['Cleaned_Review']).toarray()

In [18]:
X

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

<font color="blue">here the word by word frequency is made as an array and this is uesd to predict our model</font>

In [19]:
y = company_review['Sentiment']
y

0       1
1       0
2       0
3       0
4       0
       ..
1962    1
1963    1
1964    1
1965    1
1966    1
Name: Sentiment, Length: 1967, dtype: int64

4.2 - Training data and Test data -split

In [20]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 10)

In [21]:
len(X_train), len(X_test)

(1376, 591)

In [22]:
len(y_train), len(y_test)

(1376, 591)

4.3 - Define the Classification model (any one like NB, SVM or Random forest) train the model

In [23]:
svm_classifier = SVC(kernel='linear', C=1.0, random_state=42)
svm_classifier.fit(X_train, y_train)

SVC(kernel='linear', random_state=42)

4.4  Classification metrics analysis and plot the confusion matrix

In [24]:
y_pred = svm_classifier.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
classification_rep = classification_report(y_test, y_pred)
confusion_mat = confusion_matrix(y_test, y_pred)

print("Accuracy:", accuracy*100)
print("Classification Report:\n", classification_rep)
print("Confusion Matrix:\n", confusion_mat)

Accuracy: 81.21827411167513
Classification Report:
               precision    recall  f1-score   support

           0       0.87      0.86      0.86       410
           1       0.69      0.71      0.70       181

    accuracy                           0.81       591
   macro avg       0.78      0.78      0.78       591
weighted avg       0.81      0.81      0.81       591

Confusion Matrix:
 [[352  58]
 [ 53 128]]


### the model is correctly able to predict most of the inputs it is unable to classify certain postive and negative inputs due to hyperparameter tuning or some other preprocessing that needs to be done


### For the preprocessing and current model that is used here we are getting a accuracy of 81.2%.which is an optimal accuracy

4.5 - predict the class for your own review

In [25]:
your_review = "The redmi poco m2Pro series have a ton of complaints, such as overheating issues, bad battery and the overall review is very bad. Their sales are decreasing a lot now, as people are returning the device."
# Preprocess the review
cleaned_review = preprocess_text(your_review)
print(cleaned_review)

# Vectorize the review using the same vectorizer
your_review_vector = cv.transform([cleaned_review]).toarray()
print(your_review_vector)

# Predict the sentiment class
predicted_class = svm_classifier.predict(your_review_vector)[0]

if(predicted_class == 0):
    print("Predicted Sentiment Class for Your Review is Positive")
else:
    print("Predicted Sentiment Class for Your Review is Negative")

redmi poco m2pro seri ton complaint overh issu bad batteri overal review bad sale decreas lot peopl return devic
[[0 0 0 ... 0 0 0]]
Predicted Sentiment Class for Your Review is Negative


### The model has correctly predicted the negative review from the given input

4.6 - create feature vector using tf-idf

In [26]:
# Creating the TFIDF model
tv = TfidfVectorizer()
X_tv = tv.fit_transform(company_review['Cleaned_Review']).toarray()

In [27]:
X_train, X_test, y_train, y_test = train_test_split(X_tv, y, test_size = 0.3, random_state = 10)

4.7 Define the Classification model (any one like NB, SVM or Random forest) train the model

In [28]:
svm_classifier01 = SVC(kernel='linear', C=1.0, random_state=42)
svm_classifier01.fit(X_train, y_train)

SVC(kernel='linear', random_state=42)

4.8 - Classification metrics analysis

In [29]:
y_pred = svm_classifier01.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
classification_rep = classification_report(y_test, y_pred)
confusion_mat = confusion_matrix(y_test, y_pred)

print("Accuracy:", accuracy*100)
print("Classification Report:\n", classification_rep)
print("Confusion Matrix:\n", confusion_mat)

Accuracy: 84.4331641285956
Classification Report:
               precision    recall  f1-score   support

           0       0.86      0.93      0.89       410
           1       0.80      0.65      0.72       181

    accuracy                           0.84       591
   macro avg       0.83      0.79      0.81       591
weighted avg       0.84      0.84      0.84       591

Confusion Matrix:
 [[381  29]
 [ 63 118]]


### the model is now improving in accuracy and has reduced the misclassification.

### For the preprocessing and current model that is used here we are getting a accuracy of 84.4%.which is an optimal accuracy

### To further improve the accuracy we can also include hyperparameter tuning and try different other models such as multinomial naive bayes,bernoulli Naive bayes

4.9 - predict the class for your own review

In [30]:
# Your own review
your_review = "The redmi poco m2Pro series have a ton of complaints, such as overheating issues, bad battery and the overall review is very bad. Their sales are decreasing a lot now, as people are returning the device. Still their sales are increasing, as more and more people are buying the product as its affordable"
# Preprocess the review
cleaned_review = preprocess_text(your_review)
print(cleaned_review)

# Vectorize the review using the same vectorizer
your_review_vector = tv.transform([cleaned_review]).toarray()
# print(your_review_vector)

# Predict the sentiment class
predicted_class = svm_classifier01.predict(your_review_vector)[0]

if(predicted_class == 0):
    print("Predicted Sentiment Class for Your Review is Positive")
else:
    print("Predicted Sentiment Class for Your Review is Negative")

redmi poco m2pro seri ton complaint overh issu bad batteri overal review bad sale decreas lot peopl return devic sale increas peopl buy product afford
Predicted Sentiment Class for Your Review is Positive


### The model has correctly classified the positive review.Though this review falls to neutral side the model has learned mostly based on revenue. so it is classifying as a positive

In [31]:
your_review = "The redmi company have ton of complaints, such as overheating issues, bad battery and the overall review is very bad. Their sales are decreasing a lot now, as people are returning the device"

# Preprocess the review
cleaned_review = preprocess_text(your_review)
print(cleaned_review)

# Vectorize the review using the same vectorizer
your_review_vector = tv.transform([cleaned_review]).toarray()
# print(your_review_vector)

# Predict the sentiment class
predicted_class = svm_classifier01.predict(your_review_vector)[0]

if(predicted_class == 0):
    print("Predicted Sentiment Class for Your Review is Positive")
else:
    print("Predicted Sentiment Class for Your Review is Negative")

redmi compani ton complaint overh issu bad batteri overal review bad sale decreas lot peopl return devic
Predicted Sentiment Class for Your Review is Negative


### The model has correctly classified the negative review